cinder/volume/drivers/vmware/vmdk.py

```diff
@@ -2453,3 +2484,46 @@ def revert_to_snapshot(self, context, volume, snapshot):
             LOG.debug("Backing does not exist for volume.", resource=volume)
         else:
             self.volumeops.revert_to_snapshot(backing, snapshot.name)
+
+    def migrate_volume(self, context, volume, host):
+        """Migrate a volume to the specified host.
+
+        If the backing is not created, the dest host will create it.
+        """
+        backing = self.volumeops.get_backing(volume.name, volume.id)
+        dest_host = host['host']
+        # If the backing is not yet created, let the destination host create it
+        # In this case, migration is not necessary
+        if not backing:
+            LOG.info("There is no backing for the volume: %(volume_name)s. "
+                     "Creating it on dest_host %(dest_host)s.",
+                     {'volume_name': volume.name, 'dest_host': dest_host})
+            self._remote_api.create_backing(context, host['host'], volume)
+            return (True, None)
+
+        service_locator = self._remote_api.get_service_locator_info(context,
+                                                                    dest_host)
+        ds_info = self._remote_api.select_ds_for_volume(context, dest_host,
+                                                        volume)
+        host_ref = vim_util.get_moref(ds_info['host'], 'HostSystem')
+        rp_ref = vim_util.get_moref(ds_info['resource_pool'], 'ResourcePool')
+        ds_ref = vim_util.get_moref(ds_info['datastore'], 'Datastore')
+
+        self.volumeops.relocate_backing(backing, ds_ref, rp_ref, host_ref,
+                                        service=service_locator)
+        try:
+            self._remote_api.move_volume_backing_to_folder(
```

Comment:
Relocate spec can define folder since vSphere 6.0. Should we extend relocate_backing to make use of that property and spare the additional move step here?

Reply:
If we're to upstream this, then it has to still be compatible with lower vSpere versions I guess. This is why the existing relocate operations in the driver are followed by a move to folder.